library/ordtree.lisp

```diff
@@ -33,542 +38,508 @@
+  (define (replace t a)
+    "Returns an ordtree that has an entry `a` only if `t` already has an
+entry which is `==` to `a`.  The original entry is replaced with the given
+`a`.  If `t` doesn't have an entry `==` to `a`, `t` is returned as is."
+    (let ((rep (fn (n)
+                 (match n
+                   ((Empty)    Empty)
+                   ((N1 t1)    (make-n1 (rep t1)))
+                   ((N2 l b r)
+                    (match (<=> a b)
+                      ((LT)    (make-n2i (rep l) b r))
+                      ((EQ)    (N2 l a r))
+                      ((GT)    (make-n2i l b (rep r)))))
+                   (_ (stray-node))))))
+      (make-root (rep t))))
+
+  ;; API
+  (declare remove (Ord :elt => OrdTree :elt -> :elt -> OrdTree :elt))
+  (define (remove t a)
+    "Returns an ordtree that is the same as `t` except that the entry
+which is `==` to `a` is removed.  If `t` does not have such an entry,
+`t` is returned as is."
+    (let ((del (fn (n)
+                 (match n
+                   ((Empty) Empty)
+                   ((N1 t)  (N1 (del t)))
+                   ((N2 l b r)
+                    (match (<=> a b)
+                      ((LT) (make-n2 (del l) b r))
+                      ((EQ) (match (split-min r)
+                              ((None)               (N1 l))
+                              ((Some (Tuple a1 r1)) (make-n2 l a1 r1))))
+                      ((GT) (make-n2 l b (del r)))))
+                   (_ (stray-node))))))
+      (make-root (del t))))
+
+  (define (split-min n)
+    "Helper for removal"
+    (match n
+      ((Empty)       None)
+      ((N1 t)        (match (split-min t)
+                       ((None)              None)
+                       ((Some (Tuple a t1)) (Some (Tuple a (N1 t1))))))
+      ((N2 t1 a1 t2) (match (split-min t1)
+                       ((None)               (Some (Tuple a1 (N1 t2))))
+                       ((Some (Tuple a t11)) (Some (Tuple a
+                                                          (make-n2 t11 a1 t2))))))
+      (_ (stray-node))))
+
+  ;; API
+  (declare update (Ord :elt => OrdTree :elt -> :elt
+                       -> (Optional :elt -> (Tuple (Optional :elt) :a))
+                       -> (Tuple (OrdTree :elt) :a)))
+  (define (update t a f)
+    "Generic update.  Look for the element `a` in `t`.  If there's an entry,
+call `f` with the existing entry wrapped with Some.  If there isn't an entry,
+call `f` with None.  `f` must return a tuple of possible replacement entry,
+and an auxiliary result.
+
+If the entry doesn't exist in `t` and `f` returns `(Some elt)`, `elt` is
+inserted.  If the entry exists in `t` and `f` returns None, the element
+is removed.  If the entry exists in `t` and `f` returns `(Some elt)`, `elt`
+replaces the original entry.
+
+It is important that if `f` returns `(Some elt)`, `elt` must be `==` to
+`a`---otherwise the returned ordtree would be inconsistent.  If you use
```

Comment:
Does the function do error checking to ensure this? Or will it silently fail? If it silently fails, is it worth error checking this?